sw/host/opentitanlib/src/transport/ti50emulator/gpio.rs

```diff
@@ -1,32 +1,377 @@
 // Copyright lowRISC contributors.
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
+use std::cell::{Cell, RefCell};
+use std::fmt;
+use std::io::{Read, Write};
+use std::rc::Rc;
 
-use crate::io::gpio::{GpioPin, PinMode, PullMode};
-use crate::transport::TransportError;
-use anyhow::Result;
+use anyhow::{bail, Result};
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+use std::os::unix::net::UnixStream;
+use std::path::PathBuf;
 
-pub struct Ti50GpioPin {}
+use crate::io::emu::EmuState;
+use crate::io::gpio::{GpioError, GpioPin, PinMode, PullMode};
+use crate::transport::ti50emulator::Inner;
+use crate::transport::ti50emulator::Ti50Emulator;
+
+const GPIO_BUF_SIZE: usize = 16;
+
+/// Enumeration representing the GPIO control commands
+/// sent through the socket.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(u8)]
+enum GpioStreamCommand {
+    SetZero = b'0',
+    SetOne = b'1',
+    GetState = b'%',
+}
+
+impl From<bool> for GpioStreamCommand {
+    fn from(value: bool) -> GpioStreamCommand {
+        match value {
+            false => GpioStreamCommand::SetZero,
+            true => GpioStreamCommand::SetOne,
+        }
+    }
+}
+
+/// Enumeration that represents GPIO status updates caused
+/// by a response to a `GetState` command.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum GpioStateUpdate {
+    Value(bool),
+    PinMode(Option<PinMode>),
+    PullMode(PullMode),
+}
+
+impl TryFrom<u8> for GpioStateUpdate {
+    type Error = GpioError;
+    /// A function designed to convert the bytes sent by the GPIO stream to state changes.
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            b'0' => Ok(GpioStateUpdate::Value(false)),
+            b'1' => Ok(GpioStateUpdate::Value(true)),
+            b'z' => Ok(GpioStateUpdate::PinMode(None)),
+            b'i' => Ok(GpioStateUpdate::PinMode(Some(PinMode::Input))),
+            b'o' => Ok(GpioStateUpdate::PinMode(Some(PinMode::OpenDrain))),
+            b'b' => Ok(GpioStateUpdate::PinMode(None)),
+            b'f' => Ok(GpioStateUpdate::PinMode(None)),
+            b'x' => Ok(GpioStateUpdate::PinMode(None)),
+            b'-' => Ok(GpioStateUpdate::PullMode(PullMode::None)),
+            b'u' => Ok(GpioStateUpdate::PullMode(PullMode::PullUp)),
+            b'd' => Ok(GpioStateUpdate::PullMode(PullMode::PullDown)),
+            _ => Err(GpioError::Generic(String::from(
+                "Invalid byte during decoding GPIO stream",
+            ))),
+        }
```

Comment:
Nit: Maybe a little cleaner to collapse the redundant expressions. 

```suggestion
        Ok(match value {
            b'0' => GpioStateUpdate::Value(false),
            b'1' => GpioStateUpdate::Value(true),
            b'i' => GpioStateUpdate::PinMode(Some(PinMode::Input)),
            b'o' => GpioStateUpdate::PinMode(Some(PinMode::OpenDrain)),
            b'-' => GpioStateUpdate::PullMode(PullMode::None),
            b'u' => GpioStateUpdate::PullMode(PullMode::PullUp),
            b'd' => GpioStateUpdate::PullMode(PullMode::PullDown),
            b'z' | b'b' | b'f' | b'x' => GpioStateUpdate::PinMode(None),
            _ => return Err(GpioError::Generic(String::from(
                "Invalid byte during decoding GPIO stream",
            ))),
        })
```